### ISSUE TITLE
[Question] <title>Optimization suggestion: Implementation of refined JMX monitoring metrics

### ISSUE THREAD

**Issue by doveLin0818** [apache/hertzbeat#2908]
### Question

Current Situation Analysis (Using JMX Monitoring of Kafka as an Example):

Currently, many components' monitoring information is obtained through the JMX protocol, and the backend uses a generic method to retrieve monitoring metrics for all JMX-based monitoring logic. This leads to poor scalability of monitoring metrics and a suboptimal user experience. For example, consider the following scenarios:

First: Some metrics on the page can be merged for display. Otherwise, they may appear unfriendly and unattractive, as shown in the diagram below. Of course, users could implement this through a custom protocol, but having users define these metrics introduces a learning curve and negatively impacts the user experience. If handled by the backend, it would become easier and more visually appealing.
<img width="1383" alt="image" src="https://github.com/user-attachments/assets/e67a92c5-4eb9-4c50-8ffa-415c41c61c0a" />

Second: The generic JMX protocol-based method for retrieving metrics does not fully leverage JMX's capabilities. Since the backend must consider universality, it inevitably struggles to accommodate custom needs. For instance, the following diagram shows aggregated information for all topics in a Kafka broker. However, in daily use or within organizations, there is often more focus on the consumption status of individual topics. Let’s assume I am a HertzBeat user. To view these metrics, I first need to learn the JMX protocol for monitoring Kafka's metrics and then modify the template, as shown in the diagram below.
<img width="710" alt="image" src="https://github.com/user-attachments/assets/0a8350e2-fe7f-41dd-84ab-f3b226929a81" />
<img width="589" alt="image" src="https://github.com/user-attachments/assets/79da3c90-19e2-491a-bea9-f8d187e3bcef" />

However, HertzBeat users can only see the metrics for each topic, without knowing which specific topic the data corresponds to, as the JMX protocol does not provide a topicName metric. Displaying this requires backend implementation.
<img width="705" alt="image" src="https://github.com/user-attachments/assets/23df1ecb-0064-4ec6-bfb1-e2ea895ae260" />

Third: Companies and enterprises are also concerned about Kafka rebalance issues. This is difficult to achieve with the generic JMX monitoring method or user-customized monitoring templates. 

and many metrics cannot be implemented through the generic protocol...

In a word: The generic JMX method brings issues related to scalability. Custom development for components (such as Kafka) is needed. On the one hand, requiring users to learn JMX protocol metrics is a cost, which may lead to HertzBeat losing some competitive edge. On the other hand, the generic JMX monitoring method cannot fully harness JMX's capabilities.

If the community deems it necessary, I can attempt to customize JMX-based monitoring metrics for Kafka as a preliminary exploration.I will attempt to make Kafka's metrics more comprehensive and universal, while maintaining HertzBeat's design principles

**Comment by zhangshenghang** [comment]
thanks , @doveLin0818 

I think this is a good idea. Can you consider optimizing both Kafka Client and Kafka JMX? 
First, you can design it first, and then everyone can discuss. One thing to note: While optimizing, the newly added code should be as universal as possible.

For example, which indicators you will monitor.

**Comment by tomsun28** [comment]
+1 i think it's a good idea 👍 . It is recommended to describe your custom design here before implementing it.

**Comment by doveLin0818** [comment]
@tomsun28 @zhangshenghang  ok, I will try  to balance both practicality and HertzBeat's design principles.

**Comment by doveLin0818** [comment]
Background: Due to the limited scalability of generic JMX collection methods, it is necessary to develop a customized JMX collection solution.

Modification Process: In order to retain HertzBeat's design philosophy while supporting JMX customization, we need to first check if the current collection metrics are already registered for customization before collection. Taking Kafka's objectName=kafka.server:type=BrokerTopicMetrics,name=BytesInPerSec,topic=* as an example, we need to customize this monitoring. The process involves registering Kafka as an app in the CustomizedJmxFactory (JMX Monitoring Customization Factory) and then creating a KafkaJmxValidator to specifically handle Kafka's customization. The details are as follows:
<img width="1156" alt="image" src="https://github.com/user-attachments/assets/4019990e-c51f-4d4b-9137-0283b6fe7ef6" />

<img width="916" alt="image" src="https://github.com/user-attachments/assets/dfd913b0-ce88-47f3-b871-0eae8fd124c7" />

In summary, these components are used to determine whether the current monitoring needs to be customized. The logic for determining this is as follows:
<img width="1143" alt="image" src="https://github.com/user-attachments/assets/8f3e4d67-87ef-4912-acfa-37378a2f781c" />

If the current monitoring scenario is not effectively registered, the normal process will be followed.
If it is registered, the customization logic will be triggered, as shown in the diagram below (using Kafka as an example):
<img width="643" alt="image" src="https://github.com/user-attachments/assets/3652a9b4-3b6f-4424-9934-b2c7d89b1736" />


<img width="687" alt="image" src="https://github.com/user-attachments/assets/97a3880c-6869-4233-a81b-3aad97b8cba3" />
<img width="1121" alt="image" src="https://github.com/user-attachments/assets/4c260f72-97d4-4850-bf55-3953f6b1e9ef" />

<img width="1004" alt="image" src="https://github.com/user-attachments/assets/50011d25-3ba4-46a7-a070-7dd54baf32c1" />


Here is the preliminary demo:
<img width="2109" alt="image" src="https://github.com/user-attachments/assets/d2e244b1-53b4-483b-9966-b9d73009c036" />


This modification process effectively addresses the scalability issue of the generic JMX protocol, while also preserving the design principle that allows users to customize their protocols.

The newly added functionality has been abstracted into a factory, making it more generic. If the community agrees with this approach, I will continue to improve it. @tomsun28 @zhangshenghang

**Comment by tomsun28** [comment]
👍👍👍 Hi, Thanks! Can you provide the implementation of this code `request.getObjectInstanceSet()`? It seem is a keyone. 
<img width="538" alt="image" src="https://github.com/user-attachments/assets/2eee23c5-92a1-44ca-a764-0825d3faa2bd" />

**Comment by zhangshenghang** [comment]
@doveLin0818  Thanks. Does Kafka's JMX monitoring support all versions? The differences among different versions also need to be considered.

**Comment by doveLin0818** [comment]
> 👍👍👍 Hi, Thanks! Can you provide the implementation of this code `request.getObjectInstanceSet()`? It seem is a keyone. <img alt="image" width="538" src="https://private-user-images.githubusercontent.com/24788200/399132170-2eee23c5-92a1-44ca-a764-0825d3faa2bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0MzYzMDIsIm5iZiI6MTczNTQzNjAwMiwicGF0aCI6Ii8yNDc4ODIwMC8zOTkxMzIxNzAtMmVlZTIzYzUtOTJhMS00NGNhLWE3NjQtMDgyNWQzZmFhMmJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDAxMzMyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzZmU4NjZkMjllZWJmYzYxMTllNzg1OTM4YWUzOGMzMzU5MWE5MmIxZjkwMTZmODk2NDI4NTlmMWI4OGFmOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-dPCVpNMfiIKUuQPPqqC_FaWAkBj7dwT8gHbhOexWkk">

hi tom,this is the parameter passed from upstream, refer to the general code of JMX：
<img width="1662" alt="image" src="https://github.com/user-attachments/assets/aea6eaa5-5b21-4b0b-8227-0a93a352d0a2" />
<img width="905" alt="image" src="https://github.com/user-attachments/assets/74dba833-1f6a-4e9e-895b-fb8afdcdbc69" />

**Comment by doveLin0818** [comment]
> @doveLin0818 Thanks. Does Kafka's JMX monitoring support all versions? The differences among different versions also need to be considered.

The point you made is very important, but jmx theoretically supports all kafka versions. I will consider this issue, such as using the general jmx code as the backup logic.

**Comment by zhangshenghang** [comment]
> > @doveLin0818 Thanks. Does Kafka's JMX monitoring support all versions? The differences among different versions also need to be considered.
> 
> The point you made is very important, but jmx theoretically supports all kafka versions. I will consider this issue, such as using the general jmx code as the backup logic.

Yes, confirm to avoid different keys of JMX. I have found this problem on many services. The key formats of different versions of JMX are different.

**Comment by tomsun28** [comment]
hi, I found that the difference between old and new is the `currentObject.getKeyProperty("topic")`. Can we customize the implementation by designing jmx template protocol instead of hard coding it?
<img width="1200" alt="image" src="https://github.com/user-attachments/assets/bf5b04d7-a3c3-4f5e-96cd-be88e9ecfa60" />

![image](https://github.com/user-attachments/assets/746eb207-2f76-4c5e-952e-ab2374b17aa8)

you can find below. as the metrics `Name`, can we design a way or config to use the value of keyproperty as an metric value? so that user donot need hard code. 

<img width="663" alt="image" src="https://github.com/user-attachments/assets/055d6ef3-791c-434a-82b8-7926b0f6079b" />

**Comment by tomsun28** [comment]
Of course, your design is also universal, for other custom metrics that cannot be configured in the protocol. We recommend that support custom configuration in the template first, then hard code the way

**Comment by doveLin0818** [comment]
> hi, I found that the difference between old and new is the `currentObject.getKeyProperty("topic")`. Can we customize the implementation by designing jmx template protocol instead of hard coding it? <img alt="image" width="1200" src="https://private-user-images.githubusercontent.com/24788200/399134533-bf5b04d7-a3c3-4f5e-96cd-be88e9ecfa60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0NzcyMzYsIm5iZiI6MTczNTQ3NjkzNiwicGF0aCI6Ii8yNDc4ODIwMC8zOTkxMzQ1MzMtYmY1YjA0ZDctYTNjMy00ZjVlLTk2Y2QtYmU4OGU5ZWNmYTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDEyNTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkOWRhZWM0MDQxNjBjMDA0ZjhiNTFjNTYzNTA2OGQwOWMyNjA1MTA0YjZhMTkwYmQ1MmMzODY3YmZiOTAyNGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4aWxIgJFFljCpr2rPHvkSzZVrmsZjwkypSf-IfpRQI0">
> 
> ![image](https://private-user-images.githubusercontent.com/24788200/399134540-746eb207-2f76-4c5e-952e-ab2374b17aa8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0NzcyMzYsIm5iZiI6MTczNTQ3NjkzNiwicGF0aCI6Ii8yNDc4ODIwMC8zOTkxMzQ1NDAtNzQ2ZWIyMDctMmY3Ni00YzVlLTk1MmUtYWIyMzc0YjE3YWE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDEyNTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkZWE3OGZjYTg1OTQxNGNhNjdlNmQ0YjQ2MTljZDY1MTkxNDFhYWU0ZWFiNjM0YTk1NGExNWQ5Y2EyNjdlZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XserGmFjW-8YnIRBQjJpang9pC42JhUGPyWDfciM5Rs)
> 
> you can find below. as the metrics `Name`, can we design a way or config to use the value of keyproperty as an metric value? so that user donot need hard code.
> 
> <img alt="image" width="663" src="https://private-user-images.githubusercontent.com/24788200/399134718-055d6ef3-791c-434a-82b8-7926b0f6079b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0NzcyMzYsIm5iZiI6MTczNTQ3NjkzNiwicGF0aCI6Ii8yNDc4ODIwMC8zOTkxMzQ3MTgtMDU1ZDZlZjMtNzkxYy00MzRhLTgyYjgtNzkyNmIwZjYwNzliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDEyNTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmNDZiY2VmYTc4NjNkMjI3YWI1ZGUyNjM1YmZlODUxM2U1MGNiYzE1MDlmZGZiNDdlMmNlYjNhNGUxZDMyMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M_8YkQ3P5Orx26ff52mS8-fx7gHQlVk6w3IP9FU-MtY">

hi tom, @tomsun28 
I currently cannot obtain keyproperty through aliasFields. If I want to obtain keyproperty, I still need hard coding  to support it. However, this is only a small function for kafka. Other monitoring apps may not need keyproperty, so it is not appropriate to write it in a general template.
<img width="957" alt="image" src="https://github.com/user-attachments/assets/874de3fa-e78b-4cdc-9657-a652b66c9b96" />

And when I tried to merge ReplicaManager monitoring information through aliasFields today, I also encountered some incompatibility issues. When designing the template kafka.server:type=ReplicaManager,name=*, this MBean, I found that aliasFields did not support me to merge because their attributes are all "Value". In fact, ReplicaManager has 12 indicators, and heartbeat currently only shows two. If users want to see these 12 indicators, they need to write 12 redundant copies.
![image](https://github.com/user-attachments/assets/00dd0348-54f3-40b9-90cb-5a70f3e1aaee)
![image](https://github.com/user-attachments/assets/9dabf206-c80b-4335-904d-6af2aeb3d6f0)

So if it is necessary to fully utilize the capabilities of JMX, I currently have no good way to change it through templates, and I prefer customized development, because customization does not affect users to customize monitoring templates themselves, but only increases the workload of writing code.
